src/main/java/org/spongepowered/api/text/Transformer.java

```diff
@@ -0,0 +1,292 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.spongepowered.api.text;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * The context aware transformer interface represents a function that tries to
+ * convert the passed context data or select from them to dynamically calculate
+ * or generate content suiting the given context best. {@link Text.Placeholder}s
+ * use this interface to resolve their replacements for
+ * {@link Texts#format(Text, Function)}.
+ *
+ * @param <T> The type returned by this transformer
+ *
+ * @see Texts#format(Text, Function)
+ * @see Text.Placeholder
+ * @see Text.Placeholder#calculateReplacement(Function)
+ */
+public interface Transformer<T> {
+
+    /**
+     * Tries to transform the input context into appropriate content. Other
+     * transformations may take place after this. Implementations should be
+     * aware that sometimes their context data may not be available, in that
+     * case this method should return {@link Optional#empty()}.
+     *
+     * @param context The context function that contains the context data
+     * @return The non-null content for the given context or
+     *         {@link Optional#empty()}
+     *
+     * @see Text.Placeholder#calculateReplacement(Function)
+     */
+    Optional<T> transform(Function<String, ?> context);
```

Comment:
I also thought about this one a lot.
In the end these points made up my mind:
* You will/should never need the `keySet()` since your placeholder should only look for the keys he needs.
* You will/should never need the `values()` (Well at least I can't see a usecase)
* I found no usecase for the other `Map` methods at all. ( except `get(String)`) 
* The `Function` has the flexibility to lazily fetch the replacement. 
  * Just imagine the `arena.getPoints(player)` method is an expensive one, then you shouldn't be forced to calculate a value you may not need. And since I thought of configurable translation files here, the plugin author may not know what will be displayed there for sure.
He may limit the things he replaces, but some of them might not be used (because they could be expensive)
* I also thought about adding subtemplates (a placeholder which will print a line for each element in a list)
  * Then the list element template would be formated with mainly the current elements context with a fallback to the parents one (See example below)

Of course you can do the most things with a Map, but I just don't see a benefit.

````java
public static Transformer<Text> list(Transformer<List<?>> transformer, Text elementTemplate, Text separator, String elementKey, String indexKey) {
    checkNotNull(*);
    return context -> {
        final Optional<List<?>> optList = transformer.transform(context);
        if (optList.isPresent()) {
            List<?> list = optList.get();
            List<Text> result = new ArrayList<>();
            int index = 0;
            for (Object element : list) {
                result.add(Text.prepare(elementTemplate).with(elementKey, element).with(indexKey, index).with(context).format());
            }
            return Texts.joinOf(separator, result);
        } else {
            return Optional.empty();
        }
    }
}